### PR TITLE
ci: use .java-version and prepare to use VM with installed tools in BK

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

* Enable .java-version - that contains the current major java version
* Enable `dry-run` in the release
* Fallback to install java in Buildkite if the JAVA_HOME is not in the Buildkite runner - a follow-up to use a VM with the installed tools
* Enable maven-goal-jdk to run the JDK compatibility tests
* Use maven-goal in the remaining places that we install Java and call maven.

## Checklist

- [ ] This is an enhancement of existing features, or a new feature in existing plugins
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation
- [ ] This is a bugfix
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [x] This is something else
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)

### Why

* Stability and reliability.
* Faster builds.

### Follow-up

- VM generation should be automatically recreated if the `.java-version` changes. Likely on a timer basis (daily)

### Test

Triggering manually against this particular feature branch (for such, I use the snapshot one):
- Snapshot [build](https://buildkite.com/elastic/apm-agent-java-snapshot/builds/346) in Buildkite
- Release [build](https://buildkite.com/elastic/apm-agent-java-release/builds/27) in Buildkite

### CODE REVIEW

File: .buildkite/hooks/prepare-common.sh [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
-# JDK version is defined in two different locations, here and .github/workflows/maven-goal/action.yml
-JAVA_URL=https://jvm-catalog.elastic.co/jdk
-JAVA_HOME=$(pwd)/.openjdk17
-JAVA_PKG="$JAVA_URL/latest_openjdk_17_linux.tar.gz"
-curl -L --output /tmp/jdk.tar.gz "$JAVA_PKG"
-mkdir -p "$JAVA_HOME"
-tar --extract --file /tmp/jdk.tar.gz --directory "$JAVA_HOME" --strip-components 1
-
+# Configure the java version
+JAVA_VERSION=$(cat .java-version)
```
Comment (v1v) [inline .buildkite/hooks/prepare-common.sh:5]:
```suggestion
JAVA_VERSION=$(cat .java-version | xargs | tr -dc '[:print:]')
```

File: .ci/release.sh [hunk 2]
```diff
@@ -21,5 +21,10 @@ echo $PATH
...
 set +x
-echo "--- Deploy the release :package:"
-./mvnw -V -s .ci/settings.xml -Pgpg clean deploy -DskipTests --batch-mode | tee release.txt
+if [[ "$dry_run" == "false" ]] ; then
+  echo "--- Deploy the release :package:"
+  ./mvnw -V -s .ci/settings.xml -Pgpg clean deploy -DskipTests --batch-mode | tee release.txt
+else
+  echo "--- Deploy the release :package: (dry-run)"
+  ./mvnw -V -s .ci/settings.xml -Pgpg clean install -DskipTests --batch-mode | tee release.txt
+fi
```
Comment (v1v) [inline .ci/release.sh:30]:
```suggestion
  GOAL=deploy
else
  echo "--- Deploy the release :package: (dry-run)"
  GOAL=install
fi
./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee release.txt
```

File: .ci/snapshot.sh [hunk 3]
```diff
@@ -23,9 +23,10 @@ echo $PATH
...
-if [[ "$dry_run" == "true" ]] ; then
-  echo './mvnw -V -s .ci/settings.xml -Pgpg clean deploy -DskipTests --batch-mode'
-else
+if [[ "$dry_run" == "false" ]] ; then
+  echo "--- Deploy the snapshot :package:"
   ./mvnw -V -s .ci/settings.xml -Pgpg clean deploy -DskipTests --batch-mode | tee snapshot.txt
+else
+  echo "--- Deploy the snapshot :package: (dry-run)"
+  ./mvnw -V -s .ci/settings.xml -Pgpg clean install -DskipTests --batch-mode | tee snapshot.txt
 fi
```
Comment (v1v) [inline .ci/snapshot.sh:32]:
```suggestion
if [[ "$dry_run" == "false" ]] ; then
  echo "--- Deploy the snapshot :package:"
  GOAL=deploy
else
  echo "--- Deploy the snapshot :package: (dry-run)"
  GOAL=install
fi
./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee snapshot.txt
```

File: .github/workflows/release.yml
Comment (amannocci) [inline .github/workflows/release.yml:1]: The `dry_run` option is disabling all jobs.
Does it makes sense to add this option?

Comment (v1v) [reply .github/workflows/release.yml:1]: Good catch, there is one that should not be skipped `maven_central_deploy` but it has a dependency with `prepare_release`. I need to add specific filters per `step`.

Comment (v1v) [reply .github/workflows/release.yml:1]: ref: https://github.com/elastic/apm-agent-java/pull/3554/commits/c0a9a18e7793071358423f94687c7b3926325a5e

File: .buildkite/snapshot.yml [hunk 4]
```diff
@@ -1,5 +1,6 @@
 agents:
   provider: "gcp"
+  #image: "family/apm-agent-java-ubuntu-2204"
```
Comment (v1v) [inline .buildkite/snapshot.yml:3]: Follow-up

**Review by amannocci — DISMISSED** [review]

File: .buildkite/hooks/prepare-common.sh [hunk 5]
```diff
@@ -1,17 +1,23 @@
...
+# Fallback to install at runtime
+if [ ! -d "${JAVA_HOME}" ] ; then
+  # This should not be the case normally untless the .java-version file has been changed
+  # and the VM Image is not yet available with the latest version.
+  echo "--- Install JDK${JAVA_VERSION} :java:"
+  JAVA_URL=https://jvm-catalog.elastic.co/jdk
+  JAVA_PKG="${JAVA_URL}/latest_openjdk_${JAVA_VERSION}_linux.tar.gz"
+  curl -L --output /tmp/jdk.tar.gz "$JAVA_PKG"
+  mkdir -p "$JAVA_HOME"
+  tar --extract --file /tmp/jdk.tar.gz --directory "$JAVA_HOME" --strip-components 1
```
Comment (SylvainJuge) [inline .buildkite/hooks/prepare-common.sh:20]:
```suggestion
  curl -L --output /tmp/jdk.tar.gz "${JAVA_PKG}"
  mkdir -p "${JAVA_HOME}"
  tar --extract --file /tmp/jdk.tar.gz --directory "${JAVA_HOME}" --strip-components 1
```

File: .buildkite/hooks/prepare-common.sh [hunk 6]
```diff
@@ -1,17 +1,23 @@
...
-curl -L --output /tmp/jdk.tar.gz "$JAVA_PKG"
-mkdir -p "$JAVA_HOME"
-tar --extract --file /tmp/jdk.tar.gz --directory "$JAVA_HOME" --strip-components 1
-
+# Configure the java version
+JAVA_VERSION=$(cat .java-version | xargs | tr -dc '[:print:]')
+JAVA_HOME="${HOME}/.java/openjdk${JAVA_VERSION}"
 export JAVA_HOME
-PATH=$JAVA_HOME/bin:$PATH
+PATH="${JAVA_HOME}/bin:$PATH"
```
Comment (SylvainJuge) [inline .buildkite/hooks/prepare-common.sh:8]:
```suggestion
PATH="${JAVA_HOME}/bin:${PATH}"
```

File: .ci/release.sh [hunk 7]
```diff
@@ -21,5 +21,14 @@ echo $PATH
...
+GOAL="install"
+DRY_RUN_MSG="(dry-run)"
+# Otherwise, a RELEASE
+if [[ "$dry_run" == "false" ]] ; then
+  GOAL="deploy"
+  DRY_RUN_MSG=""
+fi
+
+echo "--- Deploy the release :package: [./mvnw $GOAL)] $DRY_RUN_MSG"
+./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee release.txt
```
Comment (SylvainJuge) [inline .ci/release.sh:34]:
```suggestion
echo "--- Deploy the release :package: [./mvnw ${GOAL})] ${DRY_RUN_MSG}"
./mvnw -V -s .ci/settings.xml -Pgpg clean ${GOAL} -DskipTests --batch-mode | tee release.txt
```

File: .ci/snapshot.sh [hunk 8]
```diff
@@ -23,9 +23,14 @@ echo $PATH
...
-echo "--- Deploy the snapshot :package:"
-if [[ "$dry_run" == "true" ]] ; then
-  echo './mvnw -V -s .ci/settings.xml -Pgpg clean deploy -DskipTests --batch-mode'
-else
-  ./mvnw -V -s .ci/settings.xml -Pgpg clean deploy -DskipTests --batch-mode | tee snapshot.txt
+# Default in dry-run mode
+GOAL="install"
+DRY_RUN_MSG="(dry-run)"
+# Otherwise, a snapshot
+if [[ "$dry_run" == "false" ]] ; then
```
Comment (SylvainJuge) [inline .ci/snapshot.sh:30]:
```suggestion
if [[ "${dry_run}" == "false" ]] ; then
```

File: .ci/snapshot.sh [hunk 9]
```diff
@@ -23,9 +23,14 @@ echo $PATH
...
+  GOAL="deploy"
+  DRY_RUN_MSG=""
 fi
+
+echo "--- Deploy the snapshot :package: [./mvnw $GOAL)] $DRY_RUN_MSG"
+./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee snapshot.txt
```
Comment (SylvainJuge) [inline .ci/snapshot.sh:36]:
```suggestion
echo "--- Deploy the snapshot :package: [./mvnw ${GOAL})] ${DRY_RUN_MSG}"
./mvnw -V -s .ci/settings.xml -Pgpg clean ${GOAL} -DskipTests --batch-mode | tee snapshot.txt
```

**Review by SylvainJuge — APPROVED** [review]

**Pull request merged**